databroker/core.py

```diff
@@ -669,7 +423,478 @@ def _external_keys(descriptor, _cache=boltons.cacheutils.LRU(max_size=500)):
+                                        "descriptor with uid {}".format(key))
+                return super().__getitem__(key)
+
+        self.denormalize = denormalize
+        self.gen = doc_gen
+        self._start = None
+        self._stop = None
+        self.descriptors = InnerDict()
+        self._events = deque()
+
+    @property
+    def start(self):
+        while self._start is None:
+            try:
+                self._get_next()
+            except StopIteration:
+                raise InvalidDocumentSequence(
+                    "stream does not contain a start?!")
+
+        return self._start
+
+    @property
+    def stop(self):
+        while self._stop is None:
+            try:
+                self._get_next()
+            except StopIteration:
+                raise InvalidDocumentSequence(
+                    "stream does not contain a start")
+
+        return self._stop
+
+    def _get_next(self):
+        self.__stash_values(*next(self.gen))
+
+    def __stash_values(self, name, doc):
+        if name == 'start':
+            if self._start is not None:
+                raise Exception("only one start allowed")
+            self._start = doc
+        elif name == 'stop':
+            if self._stop is not None:
+                raise Exception("only one stop allowed")
+            self._stop = doc
+        elif name == 'descriptor':
+            self.descriptors[doc['uid']] = doc
+        elif name == 'event':
+            self._events.append(doc)
+        else:
+            raise ValueError("{} is unknown document type".format(name))
+
+    def __denormalize(self, ev):
+        ev = dict(ev)
+        desc = ev['descriptor']
+        try:
+            ev['descriptor'] = self.descriptors[desc]
+        except StopIteration:
+            raise InvalidDocumentSequence(
+                "{} is on an event, but not in event stream".format(desc))
+        return ev
+
+    def __iter__(self):
+        gen = self.gen
+        while True:
+            while len(self._events):
+                ev = self._events.popleft()
+                if self.denormalize:
+                    ev = self.__denormalize(ev)
+                yield ev
+
+            try:
+                name, doc = next(gen)
+            except StopIteration:
+                break
+
+            if name == 'event':
+                if self.denormalize:
+                    doc = self.__denormalize(doc)
+                yield doc
+            else:
+                self.__stash_values(name, doc)
+
+
+class EventSourceShim(object):
+    '''Shim class to turn a mds object into a EventSource
+
+    This will presumably be deleted if this API makes it's way back down
+    into the implementations
+    '''
+    def __init__(self, mds, fs):
+        self.mds = mds
+        self.fs = fs
+
+    def insert(self, name, doc):
+        return self.mds.insert(name, doc)
+
+    def streams_given_header(self, header):
```

Comment:
rename to `stream_names` (*all* of these are `_given_header` -- no need to spell that out)